File: computer_science/algorithms/sorting/insertion_sort.cpp
#include <iostream>

using namespace std;

void insertion_sort(int n, int array[]) {
	int x, j;
	for (int i = 1; i < n; i++) {
		x = array[i];

		for (j = i - 1; j >= 0 && array[j] > x; j--)
			array[j+1] = array[j];

		array[j+1] = x;
	}
}

int main() {

	int array[] = { 17, 3, 2, 10 };

	insertion_sort(4, array);

	for (int i = 0; i < 4; i++) cout << array[i] << " ";

	return 0;
}

File: coding_interviews/algoexpert/insertion-sort/insertion-sort.js
// Runtime: O(N^2)
// Space: O(1)

function swap(array, i, j) {
  let aux = array[i];
  array[i] = array[j];
  array[j] = aux;
}

function insertionSort(array) {
  for (let index = 0; index < array.length; index++) {
    let toSwapIndex = index;
    while (toSwapIndex > 0 && array[toSwapIndex] < array[toSwapIndex - 1]) {
      swap(array, toSwapIndex, toSwapIndex - 1);
      toSwapIndex -= 1;
    }
  }

  return array;
}
